### ISSUE TITLE
Can I get an explanation of what linear models are doing (in plain english)?

### ISSUE THREAD

**Issue by hui-tony-zk** [STAT540-UBC/Discussion#122]
Hi @santina @marjanfarahbod 

I'm still having a lot of trouble understanding linear models.

Given a data set like so:

``` r
x <- data.frame(
  treatment = rep(c("a", "b", "c"), each = 3), 
  value = 1:9
  )
```

```
##   treatment value
## 1         a     1
## 2         a     2
## 3         a     3
## 4         b     4
## 5         b     5
## 6         b     6
## 7         c     7
## 8         c     8
## 9         c     9
```

![image](https://cloud.githubusercontent.com/assets/8253336/14031775/aeec0b0a-f1cb-11e5-8ef3-8ca900a9a359.png)

I can do a lm model:

``` r
model <- lm(value ~ treatment, x)
summary(model)
```

```
## 
## Call:
## lm(formula = value ~ treatment, data = x)
## 
## Residuals:
##    Min     1Q Median     3Q    Max 
##     -1     -1      0      1      1 
## 
## Coefficients:
##             Estimate Std. Error t value Pr(>|t|)    
## (Intercept)   2.0000     0.5774   3.464 0.013400 *  
## treatmentb    3.0000     0.8165   3.674 0.010402 *  
## treatmentc    6.0000     0.8165   7.348 0.000325 ***
## ---
## Signif. codes:  0 '***' 0.001 '**' 0.01 '*' 0.05 '.' 0.1 ' ' 1
## 
## Residual standard error: 1 on 6 degrees of freedom
## Multiple R-squared:    0.9,  Adjusted R-squared:  0.8667 
## F-statistic:    27 on 2 and 6 DF,  p-value: 0.001
```

I understand the "Estimate" represents the difference in mean of `b` or `c` compared to `a`. More specifically, `a` is encoded as `x=0`, which makes it become the `(Intercept)` (Y intercept). Then, in `b`, it is encoded as `x=1`, and the slope becomes rise/run. The rise is the mean of `b` minus the mean of `a` = 5-2 = 3, while the run is the encoded values underneath the categories, where `b` =  1 and `a` = 0 (1-0=1). The slope is therefore 3/1. The same for `c` is done.

However, what do the P-value of `b` represent? Is it the probability of observing a slope of 3 under the null hypothesis that the real slope is 0? Also, what does the p-value of `(Intercept)` represent? Is it the probability of observing an intercept of 2 under the null hypothesis that the real intercept is 0?

**Comment by marjanfarahbod** [comment]
Great and well formulated question. I already know that some students are struggling with similar things. I encourage all students to think, throw ideas and respond. We will respond later this evening or during the weekend @saramostafavi

**Comment by min110** [comment]
@hui-tony-zk thank you for bringing this here, actually it is one of my questions as well . I am still not sure if I got the right understanding, but here is how I take it from a non-stats background: 

In the lecture, we moved from one variable (focusing more on the concepts like possibility, infer the population/distribution of that variable ) to the relationship of two or more variables. but how can we study about the relationship ? 
Aspect 1: the relationship could be one to one and only one to one ( which is linear), or polynomial or even random etc. 
Aspect 2: the relationship could also be causative (regression) or not. 
Aspect 3: are they form same population or not.
I think `lm()` answered the second and the third at the same time , while the first aspect could be answered by COV and correlation .

Here, you want to know about the relationship between one categorical variable, and one continues variable . the regression line was drawn by two points (mean(variable 1), mean(variable2)); (0, intercept). 
the estimate column comes from the differences of  y values for each groups estimated from the line/model. 

About the **intercept row**, I agree with you , i think it is the comparison of reference group(a group in this case) with (0,0). Because if we don't set intercept , the regression line will be drawn by (0,0) and (mean(x),mean(y)), instead of (0,intercept) and (mean, mean)
about** b row**: it is comparison between b and a: so p value of b is under the null that b and a are two samples drawn from one population.
about c row: it is the comparison of c and a: if a and c are two samples for one population .

**Comment by hui-tony-zk** [comment]
> Here, you want to know about the relationship between one categorical variable, and one continues variable . the regression line was drawn by two points (mean(variable 1), mean(variable2)); (0, intercept). 

Not sure what you mean by this.

|treatment | mean|
|:---------|----:|
|a         |    2|
|b         |    5|
|c         |    8|

Say you want to compare `a` vs `b`. Are you saying lines are drawn from (2,5) and (0, 2)? Slope of this line = 3/2 = 1.5 (or -1.5 depending on what you're subtracting) - but this number obviously isn't the slope returned by the model.

```
coefficients(model)
```

```
## (Intercept)  treatmentb  treatmentc 
##           2           3           6
```

> the estimate column comes from the differences of y values for each groups estimated from the line/model.

The model gives you a slope and intercept, plus the associated errors from the slope and intercept. The `estimate` column is simply a reflection of the line of best fit's slope given an (0,1) encoding of categorical values. It uses no information regarding intercept, nor the associated errors.

> about **b row**: it is comparison between b and a: so p value of b is under the null that b and a are two samples drawn from one population.

I think you just described a T-test, where you use the difference in means and each group's standard deviation. Here's the output for t-tests:



```r
t.test(value ~ treatment, data = filter(x, treatment != "c"))
```

```
## 
## 	Welch Two Sample t-test
## 
## data:  value by treatment
## t = -3.6742, df = 4, p-value = 0.02131
## alternative hypothesis: true difference in means is not equal to 0
## 95 percent confidence interval:
##  -5.2669579 -0.7330421
## sample estimates:
## mean in group a mean in group b 
##               2               5
```

```r
t.test(value ~ treatment, data = filter(x, treatment != "c"), var.equal = T)
```

```
## 
## 	Two Sample t-test
## 
## data:  value by treatment
## t = -3.6742, df = 4, p-value = 0.02131
## alternative hypothesis: true difference in means is not equal to 0
## 95 percent confidence interval:
##  -5.2669579 -0.7330421
## sample estimates:
## mean in group a mean in group b 
##               2               5
```

The P-value `0.02131` is not the same as returned in the linear model `0.010402`, so that can't be right (although the T statistic is the same... what the heck).

**Comment by hui-tony-zk** [comment]
Turns out the values of `c` affects the comparison between `a` and `b`, which I guess is what you expect.


```r
x <- data.frame(
  treatment = rep(c("a", "b", "c"), each = 3), 
  value = c(1:8, 20)
  )
```

```
##   treatment value
## 1         a     1
## 2         a     2
## 3         a     3
## 4         b     4
## 5         b     5
## 6         b     6
## 7         c     7
## 8         c     8
## 9         c    20
```

![image](https://cloud.githubusercontent.com/assets/8253336/14034532/3de9855c-f1e1-11e5-8311-79457fddf4e6.png)

```r
model <- lm(value ~ treatment, x)
summary(model)
```

```
## 
## Call:
## lm(formula = value ~ treatment, data = x)
## 
## Residuals:
##    Min     1Q Median     3Q    Max 
## -4.667 -1.000  0.000  1.000  8.333 
## 
## Coefficients:
##             Estimate Std. Error t value Pr(>|t|)  
## (Intercept)    2.000      2.457   0.814   0.4467  
## treatmentb     3.000      3.475   0.863   0.4211  
## treatmentc     9.667      3.475   2.782   0.0319 *
## ---
## Signif. codes:  0 '***' 0.001 '**' 0.01 '*' 0.05 '.' 0.1 ' ' 1
## 
## Residual standard error: 4.256 on 6 degrees of freedom
## Multiple R-squared:  0.5748,	Adjusted R-squared:  0.433 
## F-statistic: 4.055 on 2 and 6 DF,  p-value: 0.07688
```

This means we can't interpret the p-value as the probability of observing slope of 3 between `a` and `b` under the null that the slope is 0. Also this further suggests that this isn't a t-test that's being performed between a and b.

**Comment by cgmay** [comment]
Just a few notes from a guy who knows very little about statistics:
1. The visualization of the model with the curves you are producing is a little misleading. Since your treatments are characters, I assume the model recognizes them as categorical (e.g. they are 3 separate drugs). Considering this, the best fit line you are drawing is essentially meaningless, as the order of categories a,b, and c is arbitrary. In this case, the estimate values are quite literally the mean differences and not the slope. If your treatment variables were numbers (e.g. different doses of the same drug) the model would treat them as continuous and the slope and other factors you have mentioned do become relevant.
2. I have no idea what the p value for the intercept means, but it seems that t tests are indeed what is being performed when looking at p values for b and c (as compared to a/intercept). You are correct in assuming that c does influence the statistic for b and a, but *I think* in a different way than you are thinking. It is not the change in slope that affects this, but rather the change in "value space" so to speak. a and b seem quite different in your first example, no overlap, good stuff. When you extend that 1 value for c, you expand the space of possible values, effectively making a and b seem more similar within the total space. But it seems weird that the p value for c has actually become larger even though you moved it further from a. This is because you have expanded the space more than you have changed the mean of c. My logic becomes faulty when you compare the lm t test between a and b (p = 0.01) to the direct t test (p = 0.02). I would have thought it would be the other way around, but I'm probably just misunderstanding some aspect of the linear model.
3. As a note on what @min110 said, this isn't exactly looking at the relationship between a categorical variable and a continuous variable. The numbers are the result, not a variable. The only variable is the treatment which in this case is categorical.

Again, this is just my interpretation, and clearly it has some significant holes.

**Comment by saramostafavi** [comment]
Hi Tony, 

You almost have it, but here are a few key things you have to take into account. 

- You have a categorical variable, so it's not intuitive to think of the estimate as the slope, but rather the 'normalized' difference (**t value**, see below) between group 2 and group 1, and then group 3 and group 1. Recall that the t value (aka, t statistics) is just (u_1 - u_2)./(std_error). So this is exactly what you are computing the t value (estimateb: (mean_group_b - mean_group_a)./(std_error). std_error is given to you in the output, but you can easily compute it (it involves a matrix inverse, look at lecture 9 slides, specifically slide 47).  In summary, what you are computing as estimates are in fact the **t values**.

- To test the significance of a t value, you need the correct degree of freedom (called residual degree of freedom--DoF). Here, you can read off from the output that the degree of freedom is 6. This depends on the number of variables in the model (actually: DoF=n_samples - variables=6). The pvalue from the t test (given the correct DoF) gives you the probability that observed t value was generated by the null model, which is centered as 0 and has specific DoF.  Note that you can reconstruct this by 2*(1-CDF_t_distribution(x=3.674,dof=6)) (this is pseudocode, translate this to R to test it in R)--you need to multiply by two , because you are considering both extreme 'tails'. 

- Obviously there are a lot of details here, and thanks for brining this up because I am sure many of you may have these questions. Linear regression by itself can be covered in a full course, so we can't cover all the details in class --- check out the books I recommended if you want to dig into this (or other topics) more deeply!

Hope this helps! Anything I missed to answer?

p.s., There is great value in being able to put complicated computational/statistical procedures into "english". In fact, I often say, in most situations if you write in english precisely enough what you want to test, you can _directly_ translate the english into an equation/formula -- but sometimes this can only be done if you expand the english concepts that you can represent  =).

**Comment by hui-tony-zk** [comment]
Hi Sarah,

I had a look on slide 47 of lecture 9, and am having difficulty interpreting the formulas - but it seems like the standard error used to calculate the t-statistic is proportional to the error estimates of all the groups (in this case, some sort of matrix multiplication of the errors associated with `a`, `b` and `c`). Is this why the standard error for all comparisons (used to calculate the t-statistic) is constant for a given linear model?

Given that this is true, does it make sense to consider the error of `c` when making a comparison between `a` and `b`? Why might this be "better" than a normal t-test?

**Comment by hui-tony-zk** [comment]
@saramostafavi, just wondering if you could help clarify why the error of `c` is important to consider when comparing `a` vs `b`?

**Comment by saramostafavi** [comment]
Hi Toni, it's not about being 'better', you should think of it as the formula for standard error once you have more than one variable (this falls out if you do the math to find the standard errors). On a related note, it is very important to keep in mind how you interpret the coefficients of a linear model once you move beyond one variable:  a regression coefficient in your linear model with more than one variable represents the mean change in your outcome (response/labels) for one unit of change in that variable while keeping the other variables in the model constant. So _what_  you include in your model is important , because obviously it will affect your estimates of the regression coefficients.  It's natural to want to avoid this responsibility ... but as an expert in your field, you _are_ the one that should make this decision. In practice, you would probably empirically try various versions of your model to get a sense of how including multiple variables would change your conclusions/results.

I don't think you can get a full picture of the details, without reading at least some chapters of the linear regression books I suggested, and importantly some linear algebra. For linear algebra, I highly highly recommend the MIT open courseware course by Gilbert Strang:  http://ocw.mit.edu/courses/mathematics/18-06-linear-algebra-spring-2010/

**Comment by louiedinh** [comment]
A more self-study friendly version of the course Sara mentions is here:
http://ocw.mit.edu/courses/mathematics/18-06sc-linear-algebra-fall-2011/
